airflow/provider.yaml.schema.json

```diff
@@ -167,6 +170,13 @@
           "python-module"
         ]
       }
+    },
+    "hook-class-names": {
```

Comment:
Above we have a 'hooks' section that accepts the python-module key. Should we take advantage of this?
https://github.com/apache/airflow/blob/fb23e61317c0173f2e4508b28fd966c776739659/airflow/provider.yaml.schema.json#L120-L143

Reply:
I wanted to explicity call out the Hook classes not modules. I think this will be faster and better than trying to find Hooks and there might some exceptions that we would have to handle (for example Hooks derived from other Hooks). however I wanted to make some check and see how many those exceptions we have and if we find that there are not many, we could indeed try to find all the classes deriving from Base Hook in the "hook" modules specified. I will take a look.

Reply:
I added a simple check now, and I think we should stay with explicitly listing the hook classes.

Just checking one amazon provider - those are Amazon's Hooks: none of them were ever "registered" in the code as "connection type to hook" nor with the customized ConnectionForm, nor in the `connection_form.js'. See below for the list of those. Similarly answering the https://github.com/apache/airflow/pull/12466#pullrequestreview-536502242  question after being a little more informed. I do not think in this case we should require all Hooks to have those extra class fields (at least not now). The use of those fields for Hooks that do not register themselves in the Core is unclear and I think we should not require the connections to have those values. 

Maybe at some point in time we want to extend the API of the BaseHook object and make it obligatory for all Hooks, but I am not sure it is something we should do now. 

IMHO - each provider should be free to choose which Hooks of that provider should be "registered" by the core for the usage in the UI and "connection to hook mapping". @ashb @mik-laj - WDYT?

Here are the Amazon hooks which do not require any special "registering" in the core (and never did):

```
'airflow.providers.amazon.aws.hooks.athena.AWSAthenaHook'
'airflow.providers.amazon.aws.hooks.athena.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.aws_dynamodb.AwsDynamoDBHook'
'airflow.providers.amazon.aws.hooks.base_aws.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.base_aws.BaseHook'
'airflow.providers.amazon.aws.hooks.batch_client.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.batch_client.AwsBatchClientHook'
'airflow.providers.amazon.aws.hooks.batch_waiters.AwsBatchClientHook'
'airflow.providers.amazon.aws.hooks.batch_waiters.AwsBatchWaitersHook'
'airflow.providers.amazon.aws.hooks.cloud_formation.AWSCloudFormationHook'
'airflow.providers.amazon.aws.hooks.cloud_formation.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.datasync.AWSDataSyncHook'
'airflow.providers.amazon.aws.hooks.datasync.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.dynamodb.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.dynamodb.AwsDynamoDBHook'
'airflow.providers.amazon.aws.hooks.ec2.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.ec2.EC2Hook'
'airflow.providers.amazon.aws.hooks.elasticache_replication_group.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.elasticache_replication_group.ElastiCacheReplicationGroupHook'
'airflow.providers.amazon.aws.hooks.emr.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.emr.EmrHook'
'airflow.providers.amazon.aws.hooks.glacier.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glacier.GlacierHook'
'airflow.providers.amazon.aws.hooks.glue.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glue.AwsGlueJobHook'
'airflow.providers.amazon.aws.hooks.glue_catalog.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.glue_catalog.AwsGlueCatalogHook'
'airflow.providers.amazon.aws.hooks.kinesis.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.kinesis.AwsFirehoseHook'
'airflow.providers.amazon.aws.hooks.lambda_function.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.lambda_function.AwsLambdaHook'
'airflow.providers.amazon.aws.hooks.logs.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.logs.AwsLogsHook'
'airflow.providers.amazon.aws.hooks.redshift.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.redshift.RedshiftHook'
'airflow.providers.amazon.aws.hooks.s3.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.s3.S3Hook'
'airflow.providers.amazon.aws.hooks.sagemaker.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sagemaker.AwsLogsHook'
'airflow.providers.amazon.aws.hooks.sagemaker.S3Hook'
'airflow.providers.amazon.aws.hooks.sagemaker.SageMakerHook'
'airflow.providers.amazon.aws.hooks.secrets_manager.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.secrets_manager.SecretsManagerHook'
'airflow.providers.amazon.aws.hooks.ses.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.ses.SESHook'
'airflow.providers.amazon.aws.hooks.sns.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sns.AwsSnsHook'
'airflow.providers.amazon.aws.hooks.sqs.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.sqs.SQSHook'
'airflow.providers.amazon.aws.hooks.step_function.AwsBaseHook'
'airflow.providers.amazon.aws.hooks.step_function.StepFunctionHook'
```

Reply:
That sounds good to me. If we need a new connection in Web UI, we have to load the module and mark it by adding this hook to this field.

Reply:
Exactly. I will add a complete documentation on "Writing your own providerr" after we merge it all, so that I do not loose time on describing things that might change.

Reply:
However, if we have a situation where some classes require a certain set of attributes and others do not, it may be a good opportunity for refacotrization. I guess this one class has two purpose now.

Reply:
What's your proposal @mik-laj ? How should we refactor it ? Also remember that it builds on previously existing fields in DBApi (I followed the same convention). 

We could use Protocol (but it's Python 3.8 https://www.python.org/dev/peps/pep-0544/) - we already backport it from typing_extensions and seems like exactly what we could use. We could have Hook also derive from DiscoverableHook (for example - I have no good name for it) and have all the hooks listed there implement it.

Reply:
For me, this is good enough for now. In the future, we can divide these classes into two separate classes. We don't really need to do this yet.

However, if you are interested, I can prepare a small snippet of code that does this.

Reply:
We can easily do it as a follow-up change :)